sql-plugin/src/main/330+/scala/org/apache/spark/sql/rapids/shims/intervalExpressions.scala

```diff
@@ -0,0 +1,497 @@
+
+import ai.rapids.cudf.{BinaryOperable, ColumnVector, DType, RoundMode, Scalar}
+import com.nvidia.spark.rapids.{Arm, GpuBinaryExpression, GpuColumnVector, GpuScalar}
+
+import org.apache.spark.sql.catalyst.expressions.{Expression, ImplicitCastInputTypes, NullIntolerant}
+import org.apache.spark.sql.types._
+
+object IntervalUtils extends Arm {
+
+  /**
+   * compute interval.toLong * num
+   */
+  def mulIntervalAndNum(interval: BinaryOperable, num: BinaryOperable): ColumnVector = {
+    interval match {
+      case intervalVector: ColumnVector =>
+        withResource(intervalVector.castTo(DType.INT64)) { intervalLong =>
+          intervalLong.mul(num)
+        }
+      case intervalScalar: Scalar =>
+        withResource(Scalar.fromLong(getLong(intervalScalar))) { intervalScalar =>
+          intervalScalar.mul(num)
+        }
+    }
+  }
+
+  /**
+   * throws exception if any value in `longCv` exceeds the int limits
+   */
+  def toIntWithCheck(longCv: ColumnVector): ColumnVector = {
+    withResource(longCv.castTo(DType.INT32)) { intResult =>
+      withResource(longCv.notEqualTo(intResult)) { notEquals =>
+        if (notEquals.any().getBoolean) {
+          throw new ArithmeticException("overflow occurs")
+        } else {
+          intResult.incRefCount()
+        }
+      }
+    }
+  }
+
+  /**
+   * Check if the long result of long * long is overflow
+   * See Math.multiplyExact:
+   * r = x * y;
+   * if (y != 0) && (r/y != x) throw exception;
+   * if(x == Long.MIN_VALUE && y == -1) throw exception
+   *
+   * @param leftLong  long cv or scalar
+   * @param rightLong long cv or scalar, will not be scalar if leftLong is scalar
+   * @param result    is the result of x * y
+   */
+  def checkMultiplyOverflow(leftLong: BinaryOperable, rightLong: BinaryOperable,
+      result: ColumnVector): Unit = {
+    leftLong match {
+      case leftVector: ColumnVector =>
+        checkMultiplyOverflowImp(rightLong, leftVector, result)
+      case _ =>
+        checkMultiplyOverflowImp(leftLong, rightLong.asInstanceOf[ColumnVector], result)
+    }
+  }
+
+  /**
+   * Check if the long result of long * long is overflow
+   * See Math.multiplyExact:
+   * r = x * y;
+   * if (y != 0) && (r/y != x) throw exception;
+   * if(x == Long.MIN_VALUE && y == -1) throw exception
+   *
+   * @param x long cv or scalar
+   * @param y long cv or scalar, will not be scalar if x is scalar
+   * @param r is the result of x * y
+   */
+  def checkMultiplyOverflowImp(x: BinaryOperable, y: ColumnVector,
+      r: ColumnVector): Unit = {
+    // check (y != 0) && (r/y != x)
+    checkMultiplyOverflowSimpleImp(x, y, r)
+
+    // check (x == Long.MIN_VALUE && y == -1)
+    x match {
+      case scalar: Scalar =>
+        if (getLong(scalar) == Long.MinValue) { // x == Long.MIN_VALUE
+          withResource(Scalar.fromLong(-1L)) { negOne =>
+            withResource(y.equalTo(negOne)) { negOneBool =>
+              if (negOneBool.any().getBoolean) { // y = -1L
```

Comment:
There is already an utils method in [BoolUtils](https://github.com/NVIDIA/spark-rapids/blob/branch-22.06/sql-plugin/src/main/scala/com/nvidia/spark/rapids/BoolUtils.scala#L48).
And the scalar returned from `negOneBool.any()` is leaked.